lotman/lotman.go

```diff
@@ -0,0 +1,655 @@
+		strPtr := *(*uintptr)(unsafe.Pointer(ptr))
+
+		// Break if the uintptr is null.
+		if strPtr == 0 {
+			break
+		}
+
+		// Create a Go string from the null-terminated string.
+		goStr := ""
+		for i := 0; ; i++ {
+			// Read the byte at the current position.
+			b := *(*byte)(unsafe.Pointer(strPtr + uintptr(i)))
+
+			// Break if the byte is null.
+			if b == 0 {
+				break
+			}
+
+			// Append the byte to the Go string.
+			goStr += string(b)
+		}
+
+		// Append the Go string to the slice.
+		goArr = append(goArr, goStr)
+
+		// Move to the next uintptr.
+		ptr += unsafe.Sizeof(uintptr(0))
+	}
+
+	return goArr
+}
+
+// Trim any buffer we get back from LotMan to the first null char
+func trimBuf(buf *[]byte) {
+	// Find the index of the first null character
+	nullIndex := bytes.IndexByte(*buf, 0)
+
+	// Trim the slice after the first null character
+	if nullIndex != -1 {
+		*buf = (*buf)[:nullIndex]
+	}
+}
+
+// Use the detected runtime to predict the location of the LotMan library.
+func getLotmanLib() string {
+	fallbackPaths := []string{
+		"/usr/lib64/libLotMan.so",
+		"/usr/local/lib64/libLotMan.so",
+		"/opt/local/lib64/libLotMan.so",
+	}
+
+	switch runtime.GOOS {
+	case "linux":
+		configuredPath := param.Lotman_LibLocation.GetString()
+		if configuredPath != "" {
+			if _, err := os.Stat(configuredPath); err == nil {
+				return configuredPath
+			}
+			log.Errorln("libLotMan.so not found in configured path, attempting to find using known fallbacks")
+		}
+
+		for _, path := range fallbackPaths {
+			if _, err := os.Stat(path); err == nil {
+				return path
+			}
+		}
+		panic("libLotMan.so not found in any of the known paths")
+	default:
+		panic(fmt.Errorf("GOOS=%s is not supported", runtime.GOOS))
+	}
+}
+
+func GetAuthorizedCallers(lotName string) (*[]string, error) {
+	success := InitLotman()
+	if !success {
+		return nil, errors.New("Failed to initialize LotMan")
+	}
+
+	// A caller is authorized if they own a parent of the lot. In the case of self-parenting lots, the owner is authorized.
+	errMsg := make([]byte, 2048)
+	cParents := unsafe.Pointer(nil)
+
+	// Get immediate parents (including self to determine rootliness). We'll use them to determine owners
+	// who are allowed to manipulate, and thus delete, the lot
+	ret := LotmanGetLotParents(lotName, false, true, &cParents, &errMsg)
+	if ret != 0 {
+		trimBuf(&errMsg)
+		return nil, errors.Errorf("Failed to determine %s's parents: %s", lotName, string(errMsg))
+	}
+
+	parents := cArrToGoArr(&cParents)
+
+	owners := []string{}
+	internalOwners := []string{}
+	for _, parent := range parents {
+		cOwners := unsafe.Pointer(nil)
+		LotmanGetLotOwners(parent, true, &cOwners, &errMsg)
+		if ret != 0 {
+			trimBuf(&errMsg)
+			return nil, errors.Errorf("Failed to determine appropriate owners of %s's parents: %s", lotName, string(errMsg))
+		}
+
+		internalOwners = append(internalOwners, cArrToGoArr(&cOwners)...)
+	}
+
+	// Deduplicate the owners
```

Comment:
Rather than making this a slice that you need to deduplicate, you can mimic a set by creating a go map from string to an empty struct and setting the owners as keys.